File: clipboard_x11.c
// Copyright 2021 The golang.design Initiative Authors.
// All rights reserved. Use of this source code is governed
// by a MIT license that can be found in the LICENSE file.
//
// Written by Changkun Ou <changkun.de>

//go:build ignore

// This is the cgo X11 clipboard implementation used only by the BSD backend,
// which #includes it textually from clipboard_bsd.c. Linux no longer compiles
// any cgo (it uses the pure-Go backend in clipboard_x11_linux.go), so this file
// carries a build constraint that never matches: it is never built as a
// standalone translation unit, only pulled in by clipboard_bsd.c.

#include <stdlib.h>
#include <stdio.h>
#include <stdint.h>
#include <string.h>
#include <dlfcn.h>
#include <X11/Xlib.h>
#include <X11/Xatom.h>

// syncStatus is a function from the Go side.
extern void syncStatus(uintptr_t handle, int status);

void *libX11;

Display* (*P_XOpenDisplay)(int);
void (*P_XCloseDisplay)(Display*);
Window (*P_XDefaultRootWindow)(Display*);
Window (*P_XCreateSimpleWindow)(Display*, Window, int, int, int, int, int, int, int);
Atom (*P_XInternAtom)(Display*, char*, int);
void (*P_XSetSelectionOwner)(Display*, Atom, Window, unsigned long);
Window (*P_XGetSelectionOwner)(Display*, Atom);
void (*P_XNextEvent)(Display*, XEvent*);
int (*P_XChangeProperty)(Display*, Window, Atom, Atom, int, int, unsigned char*, int);
void (*P_XSendEvent)(Display*, Window, int, long , XEvent*);
int (*P_XGetWindowProperty) (Display*, Window, Atom, long, long, Bool, Atom, Atom*, int*, unsigned long *, unsigned long *, unsigned char **);
void (*P_XFree) (void*);
void (*P_XDeleteProperty) (Display*, Window, Atom);
void (*P_XConvertSelection)(Display*, Atom, Atom, Atom, Window, Time);
int (*P_XSetErrorHandler)(int (*)(Display*, XErrorEvent*));
int (*P_XSync)(Display*, int);

// ignoreXError swallows asynchronous X11 protocol errors instead of letting
// Xlib's default handler terminate the whole process. On a pure Wayland
// session the X selection can be served against an invalid requestor window,
// producing a BadWindow error on X_ChangeProperty; without this handler Xlib
// would print the error and call exit(1), crashing the host program (see #61).
int ignoreXError(Display* d, XErrorEvent* e) {
	(void)d;
	(void)e;
	return 0; // the return value is ignored by Xlib
}

int initX11() {
	if (libX11) {
		return 1;
	}
	libX11 = dlopen("libX11.so", RTLD_LAZY);
    if (!libX11)
    {
        libX11 = dlopen("libX11.so.6", RTLD_LAZY);
        if (!libX11)
        {
            return 0;
        }
    }
    P_XOpenDisplay = (Display* (*)(int)) dlsym(libX11, "XOpenDisplay");
	P_XCloseDisplay = (void (*)(Display*)) dlsym(libX11, "XCloseDisplay");
	P_XDefaultRootWindow = (Window (*)(Display*)) dlsym(libX11, "XDefaultRootWindow");
	P_XCreateSimpleWindow = (Window (*)(Display*, Window, int, int, int, int, int, int, int)) dlsym(libX11, "XCreateSimpleWindow");
	P_XInternAtom = (Atom (*)(Display*, char*, int)) dlsym(libX11, "XInternAtom");
	P_XSetSelectionOwner = (void (*)(Display*, Atom, Window, unsigned long)) dlsym(libX11, "XSetSelectionOwner");
	P_XGetSelectionOwner = (Window (*)(Display*, Atom)) dlsym(libX11, "XGetSelectionOwner");
	P_XNextEvent = (void (*)(Display*, XEvent*)) dlsym(libX11, "XNextEvent");
	P_XChangeProperty = (int (*)(Display*, Window, Atom, Atom, int, int, unsigned char*, int)) dlsym(libX11, "XChangeProperty");
	P_XSendEvent = (void (*)(Display*, Window, int, long , XEvent*)) dlsym(libX11, "XSendEvent");
	P_XGetWindowProperty = (int (*)(Display*, Window, Atom, long, long, Bool, Atom, Atom*, int*, unsigned long *, unsigned long *, unsigned char **)) dlsym(libX11, "XGetWindowProperty");
	P_XFree = (void (*)(void*)) dlsym(libX11, "XFree");
	P_XDeleteProperty = (void (*)(Display*, Window, Atom)) dlsym(libX11, "XDeleteProperty");
	P_XConvertSelection = (void (*)(Display*, Atom, Atom, Atom, Window, Time)) dlsym(libX11, "XConvertSelection");
	P_XSetErrorHandler = (int (*)(int (*)(Display*, XErrorEvent*))) dlsym(libX11, "XSetErrorHandler");
	P_XSync = (int (*)(Display*, int)) dlsym(libX11, "XSync");

	// Install a handler so an asynchronous protocol error can never abort the
	// host process (see #61). This is process-global, as Xlib offers no
	// per-display protocol error handler.
	if (P_XSetErrorHandler != NULL) {
		(*P_XSetErrorHandler)(ignoreXError);
	}
	return 1;
}

// displayUnavailable remembers that opening the X11 display has already
// failed, so later calls can fail fast instead of retrying the open on
// every Write/Read/Watch. Without this, once Init fails (e.g. no DISPLAY
// or no X server) each subsequent call repeats the open attempts and
// blocks before returning the same error (see #85).
static int displayUnavailable = 0;

// openDisplay opens the X11 display, retrying a few times to tolerate a
// transient race during startup. A persistent failure is cached so that
// subsequent calls return immediately rather than retrying.
//
// Note: the fail-fast caching is not covered by a unit test. Its only
// observable effect is latency, and that only matters when XOpenDisplay
// itself is slow (e.g. a TCP timeout to an unreachable X server); with no
// DISPLAY or a missing socket it fails instantly, so there is nothing
// deterministic to assert without a timing heuristic.
Display* openDisplay() {
	if (displayUnavailable) {
		return NULL;
	}
	for (int i = 0; i < 42; i++) {
		Display* d = (*P_XOpenDisplay)(0);
		if (d != NULL) {
			return d;
		}
	}
	displayUnavailable = 1;
	return NULL;
}

int clipboard_test() {
	if (!initX11()) {
		return -1;
	}

	Display* d = openDisplay();
	if (d == NULL) {
		return -1;
	}
	(*P_XCloseDisplay)(d);
	return 0;
}

// clipboard_trigger_protocol_error deliberately issues an invalid X request to
// verify that the error handler installed by initX11 keeps the process alive.
// It returns 0 if the process survived the resulting protocol error, or a
// negative value if the X display is unavailable. Without the handler, Xlib's
// default handler would call exit(1). Exposed only for the #61 regression test.
int clipboard_trigger_protocol_error() {
	if (!initX11()) {
		return -1;
	}
	Display* d = openDisplay();
	if (d == NULL) {
		return -1;
	}
	Atom prop = (*P_XInternAtom)(d, "GOLANG_DESIGN_ERRTEST", 0);
	unsigned char val = 1;
	// 0xffffffff is not a valid window, so this yields a BadWindow protocol
	// error on X_ChangeProperty (the same opcode that crashed in #61).
	(*P_XChangeProperty)(d, (Window)0xffffffff, prop, XA_ATOM, 8, PropModeReplace, &val, 1);
	if (P_XSync != NULL) {
		(*P_XSync)(d, 0); // flush so the error is delivered to our handler
	}
	(*P_XCloseDisplay)(d);
	return 0;
}

// clipboard_write writes the given buf of size n as type typ.
// if start is provided, the value of start will be changed to 1 to indicate
// if the write is availiable for reading.
int clipboard_write(char *typ, unsigned char *buf, size_t n, uintptr_t handle) {
	if (!initX11()) {
        syncStatus(handle, -1);
		return -1;
	}

    Display* d = openDisplay();
    if (d == NULL) {
        syncStatus(handle, -1);
        return -1;
    }
    Window w = (*P_XCreateSimpleWindow)(d, (*P_XDefaultRootWindow)(d), 0, 0, 1, 1, 0, 0, 0);

    // Use False because these may not available for the first time.
    Atom sel         = (*P_XInternAtom)(d, "CLIPBOARD", 0);
    Atom atomString  = (*P_XInternAtom)(d, "UTF8_STRING", 0);
    Atom atomImage   = (*P_XInternAtom)(d, "image/png", 0);
    Atom targetsAtom = (*P_XInternAtom)(d, "TARGETS", 0);

    // Use True to makesure the requested type is a valid type.
    Atom target = (*P_XInternAtom)(d, typ, 1);
    if (target == None) {
        (*P_XCloseDisplay)(d);
        syncStatus(handle, -2);
        return -2;
    }

    (*P_XSetSelectionOwner)(d, sel, w, CurrentTime);
    if ((*P_XGetSelectionOwner)(d, sel) != w) {
        (*P_XCloseDisplay)(d);
        syncStatus(handle, -3);
        return -3;
    }

    XEvent event;
    XSelectionRequestEvent* xsr;
    int notified = 0;
    for (;;) {
        if (notified == 0) {
            syncStatus(handle, 1); // notify Go side
            notified = 1;
        }

        (*P_XNextEvent)(d, &event);
        switch (event.type) {
        case SelectionClear:
            // For debugging:
            // printf("x11write: lost ownership of clipboard selection.\n");
            // fflush(stdout);
            (*P_XCloseDisplay)(d);
            return 0;
        case SelectionNotify:
            // For debugging:
            // printf("x11write: notify.\n");
            // fflush(stdout);
            break;
        case SelectionRequest:
            if (event.xselectionrequest.selection != sel) {
                break;
            }

            XSelectionRequestEvent * xsr = &event.xselectionrequest;
            XSelectionEvent ev = {0};
            int R = 0;

            ev.type      = SelectionNotify;
            ev.display   = xsr->display;
            ev.requestor = xsr->requestor;
            ev.selection = xsr->selection;
            ev.time      = xsr->time;
            ev.target    = xsr->target;
            ev.property  = xsr->property;

            if (ev.target == atomString && ev.target == target) {
                R = (*P_XChangeProperty)(ev.display, ev.requestor, ev.property,
                    atomString, 8, PropModeReplace, buf, n);
            } else if (ev.target == atomImage && ev.target == target) {
                R = (*P_XChangeProperty)(ev.display, ev.requestor, ev.property,
                    atomImage, 8, PropModeReplace, buf, n);
            } else if (ev.target == targetsAtom) {
                // Reply atoms for supported targets, other clients should
                // request the clipboard again and obtain the data if their
                // implementation is correct.
                Atom targets[] = { targetsAtom, target };
                R = (*P_XChangeProperty)(ev.display, ev.requestor, ev.property,
                    XA_ATOM, 32, PropModeReplace,
                    (unsigned char *)&targets, sizeof(targets)/sizeof(Atom));
            } else {
                ev.property = None;
            }

            if ((R & 2) == 0) (*P_XSendEvent)(d, ev.requestor, 0, 0, (XEvent *)&ev);
            break;
        }
    }
}

// read_data reads the property of a selection if the target atom matches
// the actual atom.
unsigned long read_data(XSelectionEvent *sev, Atom sel, Atom prop, Atom target, char **buf) {
	if (!initX11()) {
		return -1;
	}

    unsigned char *data;
    Atom actual;
    int format;
    unsigned long n    = 0;
    unsigned long size = 0;
    if (sev->property == None || sev->selection != sel || sev->property != prop) {
        return 0;
    }

    int ret = (*P_XGetWindowProperty)(sev->display, sev->requestor, sev->property,
        0L, (~0L), 0, AnyPropertyType, &actual, &format, &size, &n, &data);
    if (ret != Success) {
        return 0;
    }

    if (actual == target && buf != NULL) {
        *buf = (char *)malloc(size * sizeof(char));
        memcpy(*buf, data, size*sizeof(char));
    }
    (*P_XFree)(data);
    (*P_XDeleteProperty)(sev->display, sev->requestor, sev->property);
    return size * sizeof(char);
}

// clipboard_read reads the clipboard selection in given format typ.
// the read bytes are written into buf and returns the size of the buffer.
//
// The caller of this function should responsible for the free of the buf.
unsigned long clipboard_read(char* typ, char **buf) {
	if (!initX11()) {
		return -1;
	}

    Display* d = openDisplay();
    if (d == NULL) {
        return -1;
    }

    Window w = (*P_XCreateSimpleWindow)(d, (*P_XDefaultRootWindow)(d), 0, 0, 1, 1, 0, 0, 0);

    // Use False because these may not available for the first time.
    Atom sel  = (*P_XInternAtom)(d, "CLIPBOARD", False);
    Atom prop = (*P_XInternAtom)(d, "GOLANG_DESIGN_DATA", False);

    // Use True to makesure the requested type is a valid type.
    Atom target = (*P_XInternAtom)(d, typ, True);
    if (target == None) {
        (*P_XCloseDisplay)(d);
        return -2;
    }

    (*P_XConvertSelection)(d, sel, target, prop, w, CurrentTime);
    XEvent event;
    for (;;) {
        (*P_XNextEvent)(d, &event);
        if (event.type != SelectionNotify) continue;
        break;
    }
    unsigned long n = read_data((XSelectionEvent *)&event.xselection, sel, prop, target, buf);
    (*P_XCloseDisplay)(d);
    return n;
}

File: clipboard_x11.go
// Copyright 2026 The golang.design Initiative Authors.
// All rights reserved. Use of this source code is governed
// by a MIT license that can be found in the LICENSE file.
//
// Written by Changkun Ou <changkun.de>

//go:build (linux || freebsd || openbsd || netbsd) && !android

package clipboard

// Pure-Go X11 CLIPBOARD selection backend, shared by Linux and the BSDs. It
// speaks the X11 wire protocol (encoded/decoded by internal/x11wire) directly
// over the display socket, so it needs no Cgo, no libX11 headers at build time,
// and no libX11.so at runtime.

import (
	"bufio"
	"encoding/binary"
	"net"
	"os"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"
	"time"

	x11wire "golang.design/x/x11"
)

// x11ReadTimeout bounds a single Read so a missing SelectionNotify (e.g. an
// owner that never answers) surfaces as an error instead of hanging.
const x11ReadTimeout = 5 * time.Second

// x11conn is a live connection to the X server.
type x11conn struct {
	c     net.Conn
	r     *bufio.Reader
	ids   *x11wire.IDGen
	root  uint32
	win   uint32
	atoms map[string]uint32
	seq   uint16 // last request sequence number sent (server counts from 1)
}

// send writes a request and returns the sequence number the server assigns it,
// so a later reply can be matched by sequence.
func (x *x11conn) send(req []byte) (uint16, error) {
	if _, err := x.c.Write(req); err != nil {
		return 0, err
	}
	x.seq++
	return x.seq, nil
}

// reply reads packets until it returns the reply matching seq. Events and
// packets for other sequence numbers are discarded; in particular, asynchronous
// X11 errors caused by earlier fire-and-forget requests (e.g. a requestor
// window that has gone away) are dropped rather than mistaken for this reply —
// the same reason a protocol error cannot crash the process (#61). An error
// whose sequence matches seq means this request itself failed.
func (x *x11conn) reply(seq uint16) (x11wire.Packet, error) {
	for {
		p, err := x11wire.ReadPacket(x.r)
		if err != nil {
			return x11wire.Packet{}, err
		}
		if p.IsEvent() || p.Sequence() != seq {
			continue
		}
		if p.IsError() {
			return x11wire.Packet{}, errUnavailable
		}
		return p, nil
	}
}

// x11Dial connects to the display socket, trying the filesystem unix socket
// first, then the Linux abstract socket, then TCP.
func x11Dial(d x11wire.Display) (net.Conn, error) {
	if d.Net == "unix" {
		c, err := net.Dial("unix", d.Addr)
		if err == nil || runtime.GOOS != "linux" {
			// The abstract-socket fallback below is Linux-only; the BSDs have
			// no abstract unix namespace, so there is nothing more to try.
			return c, err
		}
		return net.Dial("unix", "@/tmp/.X11-unix/X"+strconv.Itoa(d.Num))
	}
	return net.Dial(d.Net, d.Addr)
}

// loadCookie reads the MIT-MAGIC-COOKIE-1 for the given display from
// $XAUTHORITY (or ~/.Xauthority). It returns ("", nil) when none applies, in
// which case we connect without authorization.
func loadCookie(displayNum int) (string, []byte) {
	path := os.Getenv("XAUTHORITY")
	if path == "" {
		home, err := os.UserHomeDir()
		if err != nil {
			return "", nil
		}
		path = filepath.Join(home, ".Xauthority")
	}
	b, err := os.ReadFile(path)
	if err != nil {
		return "", nil
	}
	entries, err := x11wire.ParseXauthority(b)
	if err != nil {
		return "", nil
	}
	host, _ := os.Hostname()
	return x11wire.ChooseCookie(entries, displayNum, host)
}

// x11Handshake performs the connection setup over an already-dialed socket.
func x11Handshake(c net.Conn, name string, data []byte) (*bufio.Reader, x11wire.Setup, error) {
	if _, err := c.Write(x11wire.SetupRequest(name, data)); err != nil {
		return nil, x11wire.Setup{}, err
	}
	r := bufio.NewReader(c)
	s, err := x11wire.ReadSetup(r)
	return r, s, err
}

// x11DialDisplay dials the display and authenticates, returning a connection
// with no window yet. Splitting this from window creation lets x11Test check
// reachability without churning a window resource on the server.
//
// The connection setup is retried a few times: under rapid connect/disconnect
// churn the X server can reset a connection mid-handshake ("connection reset by
// peer"). The original cgo backend tolerated the same transient by retrying
// XOpenDisplay many times.
func x11DialDisplay() (*x11conn, error) {
	d, err := x11wire.ParseDisplay(os.Getenv("DISPLAY"))
	if err != nil {
		return nil, errUnavailable
	}
	name, data := loadCookie(d.Num)

	for range 10 {
		x, err := x11dialOnce(d, name, data)
		if err == nil {
			return x, nil
		}
		time.Sleep(5 * time.Millisecond)
	}
	return nil, errUnavailable
}

// x11dialOnce makes one connection-setup attempt, falling back to no
// authorization (common for local unix sockets) if the cookie is rejected.
func x11dialOnce(d x11wire.Display, name string, data []byte) (*x11conn, error) {
	c, err := x11Dial(d)
	if err != nil {
		return nil, err
	}
	r, setup, err := x11Handshake(c, name, data)
	if err != nil {
		c.Close()
		if name == "" {
			return nil, err
		}
		if c, err = x11Dial(d); err != nil {
			return nil, err
		}
		if r, setup, err = x11Handshake(c, "", nil); err != nil {
			c.Close()
			return nil, err
		}
	}
	return &x11conn{c: c, r: r, ids: x11wire.NewIDGen(setup), root: setup.Root, atoms: map[string]uint32{}}, nil
}

// x11Connect dials, authenticates, and creates the 1x1 window used to own or
// request the selection.
func x11Connect() (*x11conn, error) {
	x, err := x11DialDisplay()
	if err != nil {
		return nil, err
	}
	if err := x.createWindow(); err != nil {
		x.Close()
		return nil, err
	}
	return x, nil
}

// createWindow creates x.win and confirms it succeeded, retrying with a fresh
// resource id if the server rejects the id. The X server may hand a new
// connection the same resource-id base a just-closed connection used; creating
// a window with an id the server has not yet reaped yields BadIDChoice. A
// GetInputFocus sync surfaces that asynchronous error so we can retry.
func (x *x11conn) createWindow() error {
	for range 8 {
		x.win = x.ids.Next()
		cwSeq, err := x.send(x11wire.CreateWindow(x.win, x.root))
		if err != nil {
			return errUnavailable
		}
		syncSeq, err := x.send(x11wire.GetInputFocus())
		if err != nil {
			return errUnavailable
		}
		created := true
		for {
			p, err := x11wire.ReadPacket(x.r)
			if err != nil {
				return errUnavailable
			}
			if p.IsEvent() {
				continue
			}
			if p.IsError() && p.Sequence() == cwSeq {
				created = false // the id was rejected; try the next one
				continue
			}
			if p.Sequence() == syncSeq {
				break // CreateWindow result has been flushed
			}
		}
		if created {
			return nil
		}
	}
	return errUnavailable
}

func (x *x11conn) Close() error { return x.c.Close() }

// intern resolves an atom name to its id, caching the result.
func (x *x11conn) intern(name string) (uint32, error) {
	if a, ok := x.atoms[name]; ok {
		return a, nil
	}
	seq, err := x.send(x11wire.InternAtom(name, false))
	if err != nil {
		return 0, err
	}
	p, err := x.reply(seq)
	if err != nil {
		return 0, err
	}
	a := p.Atom()
	x.atoms[name] = a
	return a, nil
}

// x11Test verifies that the X server can be reached. Used by initialize. It
// only dials and authenticates; it does not create a window, to avoid churning
// a server resource on every Init.
func x11Test() error {
	c, err := x11DialDisplay()
	if err != nil {
		return err
	}
	c.Close()
	return nil
}

// x11Read reads the CLIPBOARD selection in the given target format.
func x11Read(target string) ([]byte, error) {
	x, err := x11Connect()
	if err != nil {
		return nil, errUnavailable
	}
	defer x.Close()
	x.c.SetReadDeadline(time.Now().Add(x11ReadTimeout))

	sel, e1 := x.intern("CLIPBOARD")
	prop, e2 := x.intern("GOLANG_DESIGN_DATA")
	tgt, e3 := x.intern(target)
	if e1 != nil || e2 != nil || e3 != nil {
		return nil, errUnavailable
	}

	if _, err := x.send(x11wire.ConvertSelection(x.win, sel, tgt, prop, x11wire.CurrentTime)); err != nil {
		return nil, errUnavailable
	}
	for {
		p, err := x11wire.NextEvent(x.r)
		if err != nil {
			return nil, errUnavailable
		}
		if p.EventCode() != x11wire.EventSelectionNotify {
			continue
		}
		if p.SelectionNotify().Property == x11wire.None {
			return nil, nil // nothing available in this format
		}
		break
	}

	gseq, err := x.send(x11wire.GetProperty(true, x.win, prop, 0, 0, 0xffffffff))
	if err != nil {
		return nil, errUnavailable
	}
	rp, err := x.reply(gseq)
	if err != nil {
		return nil, errUnavailable
	}
	v := rp.PropertyValue()
	if len(v) == 0 {
		// An empty property reads back as nil, matching the historical cgo
		// behavior (and Read's documented "returns nil when absent").
		return nil, nil
	}
	return v, nil
}

// atomName resolves an atom id to its string name via GetAtomName.
func (x *x11conn) atomName(atom uint32) (string, error) {
	seq, err := x.send(x11wire.GetAtomName(atom))
	if err != nil {
		return "", err
	}
	p, err := x.reply(seq)
	if err != nil {
		return "", err
	}
	return p.AtomName(), nil
}

// x11Targets returns the target names the current CLIPBOARD selection advertises
// (via the TARGETS target), or nil if the clipboard is empty. The TARGETS
// property is a list of 4-byte atom ids, each resolved back to its name.
func x11Targets() ([]string, error) {
	x, err := x11Connect()
	if err != nil {
		return nil, errUnavailable
	}
	defer x.Close()
	x.c.SetReadDeadline(time.Now().Add(x11ReadTimeout))

	sel, e1 := x.intern("CLIPBOARD")
	prop, e2 := x.intern("GOLANG_DESIGN_DATA")
	tgts, e3 := x.intern("TARGETS")
	if e1 != nil || e2 != nil || e3 != nil {
		return nil, errUnavailable
	}

	if _, err := x.send(x11wire.ConvertSelection(x.win, sel, tgts, prop, x11wire.CurrentTime)); err != nil {
		return nil, errUnavailable
	}
	for {
		p, err := x11wire.NextEvent(x.r)
		if err != nil {
			return nil, errUnavailable
		}
		if p.EventCode() != x11wire.EventSelectionNotify {
			continue
		}
		if p.SelectionNotify().Property == x11wire.None {
			return nil, nil // empty clipboard
		}
		break
	}

	gseq, err := x.send(x11wire.GetProperty(true, x.win, prop, 0, 0, 0xffffffff))
	if err != nil {
		return nil, errUnavailable
	}
	rp, err := x.reply(gseq)
	if err != nil {
		return nil, errUnavailable
	}

	atoms := rp.PropertyValue() // ATOM list, 4 bytes each
	names := make([]string, 0, len(atoms)/4)
	for i := 0; i+4 <= len(atoms); i += 4 {
		name, err := x.atomName(binary.LittleEndian.Uint32(atoms[i:]))
		if err != nil || name == "" {
			continue
		}
		names = append(names, name)
	}
	return names, nil
}

// x11EnumerateFormats maps the advertised TARGETS to Format tokens, registering
// custom MIME types on demand. Shared by the Linux and BSD backends.
func x11EnumerateFormats() []Format {
	names, err := x11Targets()
	if err != nil {
		return nil
	}
	out := make([]Format, 0, len(names))
	for _, n := range names {
		if f, ok := x11FormatForTarget(n); ok {
			out = append(out, f)
		}
	}
	return out
}

// x11FormatForTarget maps an X11 target/atom name to a Format: the common text
// atoms to FmtText, image/png to FmtImage, any other MIME-shaped name (one that
// contains '/') to a registered custom format. X11 meta-targets such as TARGETS,
// MULTIPLE and TIMESTAMP have no '/', so they are ignored.
func x11FormatForTarget(name string) (Format, bool) {
	switch name {
	case "UTF8_STRING", "STRING", "TEXT", "text/plain", "text/plain;charset=utf-8":
		return FmtText, true
	case "image/png":
		return FmtImage, true
	}
	if strings.Contains(name, "/") {
		return Register(name), true
	}
	return 0, false
}

// x11Write takes ownership of the CLIPBOARD selection and serves its content to
// requestors until ownership is lost (another writer overwrites it). The
// returned channel is closed on that loss, matching the documented contract.
func x11Write(target string, buf []byte) (<-chan struct{}, error) {
	x, err := x11Connect()
	if err != nil {
		return nil, errUnavailable
	}

	sel, e1 := x.intern("CLIPBOARD")
	targets, e2 := x.intern("TARGETS")
	tgt, e3 := x.intern(target)
	if e1 != nil || e2 != nil || e3 != nil {
		x.Close()
		return nil, errUnavailable
	}

	if _, err := x.send(x11wire.SetSelectionOwner(x.win, sel, x11wire.CurrentTime)); err != nil {
		x.Close()
		return nil, errUnavailable
	}
	gseq, err := x.send(x11wire.GetSelectionOwner(sel))
	if err != nil {
		x.Close()
		return nil, errUnavailable
	}
	p, err := x.reply(gseq)
	if err != nil || p.SelectionOwner() != x.win {
		x.Close()
		return nil, errUnavailable
	}

	done := make(chan struct{}, 1)
	go func() {
		defer x.Close()
		x.serveSelection(sel, targets, tgt, buf)
		done <- struct{}{}
		close(done)
	}()
	return done, nil
}

// serveSelection runs the owner event loop, answering selection requests until
// a SelectionClear (ownership lost) or a connection error.
func (x *x11conn) serveSelection(sel, targets, tgt uint32, buf []byte) {
	for {
		p, err := x11wire.NextEvent(x.r)
		if err != nil {
			return
		}
		switch p.EventCode() {
		case x11wire.EventSelectionClear:
			return
		case x11wire.EventSelectionRequest:
			req := p.SelectionRequest()
			if req.Selection == sel {
				x.answerSelectionRequest(req, targets, tgt, buf)
			}
		}
	}
}

// answerSelectionRequest replies to a single SelectionRequest: it serves the
// data for our target, the supported list for TARGETS, or refuses otherwise,
// then notifies the requestor.
func (x *x11conn) answerSelectionRequest(req x11wire.SelectionRequestEvent, targets, tgt uint32, buf []byte) {
	notify := x11wire.SelectionNotify{
		Time:      req.Time,
		Requestor: req.Requestor,
		Selection: req.Selection,
		Target:    req.Target,
		Property:  req.Property,
	}
	switch req.Target {
	case tgt:
		x.send(x11wire.ChangeProperty(req.Requestor, req.Property, tgt, 8, buf))
	case targets:
		// Advertise the supported targets so correct clients re-request the
		// data in a format we serve (#60).
		x.send(x11wire.ChangeProperty(req.Requestor, req.Property,
			x11wire.AtomATOM, 32, x11wire.AtomList(targets, tgt)))
	default:
		notify.Property = x11wire.None // refuse unsupported target
	}
	x.send(x11wire.SendSelectionNotify(notify))
}
